### PR TITLE
THRIFT-3486 - Java generated `getFieldValue` is incompatible with `setFieldValue` for binary values.

### PR DESCRIPTION
**Pull request by bgould**
For binary fields, the compiler generates Java getter/setter methods for primitive byte arrays (`byte[]`) as well as `java.nio.ByteBuffer` objects.  However, `setFieldValue` only accepts a `ByteBuffer`, and throws a `ClassCastException` if a `byte[]` is passed as an argument. To further confuse things, `getFieldValue` returns a `byte[]`. By allowing either a `ByteBuffer` or a `byte[]` as input, `setFieldValue` could exhibit less confusing behavior for end users.

These changes add an `instanceof` check for binary fields in the generated `setFieldValue()` method so that it accepts either a `byte[]` or a `ByteBuffer`.  There is also an update to the Java library's test suite to test for this functionality.

### CODE REVIEW

File: lib/java/test/org/apache/thrift/TestStruct.java [hunk 1]
```diff
@@ -309,11 +309,33 @@ public void testToString() throws Exception {
...
+    o.setReq_bin(testString.getBytes());
+    assertNotNull(o.getReq_bin());
+
+    o.setFieldValue(JavaTestHelper._Fields.REQ_BIN, null);
+    assertNull(o.getReq_bin());
+
+    o.setFieldValue(JavaTestHelper._Fields.REQ_BIN, testString.getBytes());
+    assertNotNull(o.getReq_bin());
+
+    o.setFieldValue(JavaTestHelper._Fields.REQ_BIN, null);
```
Comment (Gospodzin) [inline lib/java/test/org/apache/thrift/TestStruct.java]: Does it make sense to duplicate this piece of code?

File: lib/java/test/org/apache/thrift/TestStruct.java [hunk 2]
```diff
@@ -309,11 +309,33 @@ public void testToString() throws Exception {
...
-    JavaTestHelper o = new JavaTestHelper();
+    
+    final String testString = "testBytesBufferFeatures";
+    final JavaTestHelper o = new JavaTestHelper();
+
     o.setReq_bin((ByteBuffer)null);
     assertNull(o.getReq_bin());
+
+    o.setReq_bin(ByteBuffer.wrap(testString.getBytes()));
+    assertNotNull(o.getReq_bin());
```
Comment (Gospodzin) [inline lib/java/test/org/apache/thrift/TestStruct.java]: Wouldn't it make more sense to use `assertArrayEquals(testString.getBytes(), o.getReq_bin())` (same at line 326)?

File: lib/java/test/org/apache/thrift/TestStruct.java (hunk 1 above)
Comment (bgould) [inline lib/java/test/org/apache/thrift/TestStruct.java]: Just wanted to make sure the field was null before proceeding before testing the next "set" operation

File: lib/java/test/org/apache/thrift/TestStruct.java (hunk 2 above)
Comment (bgould) [inline lib/java/test/org/apache/thrift/TestStruct.java]: Yes, couldn't hurt.  Really I was just trying to test to make sure it doesn't trigger a ClassCastException; but your suggestion seems logical.

File: lib/java/test/org/apache/thrift/TestStruct.java (hunk 1 above)
Comment (Gospodzin) [inline lib/java/test/org/apache/thrift/TestStruct.java]: seems logical ;)

**Pull request closed without merging**